programs/bpf_loader/src/serialization.rs

```diff
@@ -247,9 +260,14 @@ pub fn serialize_parameters_aligned(
 pub fn deserialize_parameters_aligned(
     keyed_accounts: &[KeyedAccount],
     buffer: &[u8],
+    account_lengths: &[usize],
```

Comment:
This is neccessary because a CPI call will change the length of the account within this call?

Reply:
Yes, that was the missing puzzle piece I searched for the entire day.
The hidden purpose of the additional temporary account structs.

Because once the CPI calls recycle the existing account structs, it overwrites the length and so the deserialize of the caller program afterwards parses and writes back garbage.